posts/2022-12-14-android-ndk-update.md

```diff
@@ -0,0 +1,25 @@
+---
+layout: post
+title: "Updating the Android NDK in Rust <version>"
+author: Android Platform Team
+description: "Modernizing Android support in Rust"
+---
+
+We are pleased to announce that Android platform support in Rust will be
+modernized in Rust <version> as we update the target NDK from r17 to r25.  As a
+consequence the minimum supported API level will increase from 15 (Ice Cream
+Sandwich) to 19 (KitKat).
+
+In NDK r23 Android switched to using LLVM's `libunwind` for all architectures.
+This meant that
+1. If a project were to target NDK r23 or newer with previous versions of Rust
+   [a workaround](https://github.com/rust-lang/rust/pull/85806#issuecomment-1096266946)
+   would be required to redirect attempts to link against `libgcc` to instead
+   link against `libunwind`.  Following this update this workaround will no
+   longer be necessary.
+2. If a project uses NDK r22 or older it will need to be updated to use r23 or
+   newer.
+
+Going forward the Android platform will target the most recent LTS NDK, allowing
```

Comment:
Have we discussed how quickly this bump will happen? I think at minimum we probably bump when it releases on nightly, which gives folks 6-12 weeks to update if needed. It seems reasonable to guarantee 10-12 weeks, but no strong feelings there.

I do think we should add a sentence here like "this will be mentioned in release notes", and if possible provide guidance on how often it happens (yearly?).

Reply:
I'd like to set up some kind of automation somewhere that tests nightly against NDK betas.  I'm not sure if that would give us enough confidence to bump the NDK version the Rust release immediately after the NDK release, but it would at least allow us to file potential issues during the beta process.

If that strategy wouldn't work then I think adding the new NDK to nightly and let it go through the normal process is reasonable.

I believe NDK LTS releases will occur at most once a year going forward.

Reply:
> confidence to bump the NDK version the Rust release immediately after the NDK release

I don't think we would ever want to be backporting the NDK bump, so unless you mean that we would bump to a not yet(?) released version, I don't see this happening regardless of how much testing it received.

We do bump to not-yet-released versions for LLVM, so that might be feasible, but I imagine that the community pain with LLVM (which the vast majority of Rust users aren't building, and if they are, we're not forcing them to immediately jump to a different branch on our switch) is significantly lower than with NDKs, at least from what I understand.

Reply:
My idea was to have a special CI configuration that builds and tests nightly with the Android NDK beta.  The goal being to catch any issues before the actual nightly builds are moved to newly released versions of the NDK.

I'm not suggesting that Rust releases be built against beta NDKs.

Reply:
That doesn't seem unreasonable, though to some extent depends on how often that builder would break and whether it would take a significant chunk of time.

I think part of the confusion might be that to be honest, bumping the version to the new NDK version directly after that releases on nightly Rust isn't something I think we need *confidence* for -- we either have policy that says we do that or not, but it's not really based on knowing whether such a bump will land. It's obviously gated on it landing (and bors approving that), but not more than that.

Reply:
With that setup do you think it would be possible to skip the nightly branch and move the NDK updates directly into the beta branch?

Reply:
No, we don't backport or directly land changes in the beta branch for anything except to fix regressions - I don't think this would qualify.